src/Utilities/Tags.jl

```diff
@@ -0,0 +1,120 @@
+])
+const INLINE_ELEMENTS = Set([
+    :a, :abbr, :acronym, :b, :bdo, :big, :br, :button, :cite, :code, :dfn, :em,
+    :i, :img, :input, :kbd, :label, :map, :object, :q, :samp, :script, :select,
+    :small, :span, :strong, :sub, :sup, :textarea, :time, :tt, :var,
+])
+const VOID_ELEMENTS = Set([
+    :area, :base, :br, :col, :command, :embed, :hr, :img, :input, :keygen,
+    :link, :meta, :param, :source, :track, :wbr,
+])
+const ALL_ELEMENTS = union(BLOCK_ELEMENTS, INLINE_ELEMENTS, VOID_ELEMENTS)
+
+const EMPTY_STRING = ""
+const TEXT = Symbol(EMPTY_STRING)
+
+immutable Tag
+    name :: Symbol
+end
+
+Base.show(io::IO, t::Tag) = print(io, "<", t.name, ">")
+
+macro tags(args...) esc(tags(args)) end
+tags(s) = :(const ($(s...),) = $(map(Tag, s)))
+
+typealias Attributes Vector{Pair{Symbol, String}}
+
+immutable Node
+    name :: Symbol
+    text :: String
+    attr :: Attributes
+    data :: Vector{Node}
+
+    Node(name::Symbol, attr::Attributes, data::Vector{Node}) = new(name, EMPTY_STRING, attr, data)
+    Node(name::Symbol, text::AbstractString) = (@assert(name === TEXT); new(name, text))
+end
+
+@compat (t::Tag)(args...) = Node(t.name, Attributes(), data(args))
+@compat (n::Node)(args...) = Node(n.name, n.attr, data(args))
+Base.getindex(t::Tag, args...) = Node(t.name, attr(args), Node[])
+Base.getindex(n::Node, args...) = Node(n.name, attr(args), n.data)
+
+data(args) = flatten!(nodes!, Node[], args)
+attr(args) = flatten!(attributes!, Attributes(), args)
+
+typealias Atom Union{AbstractString, Node, Pair{Symbol, String}, Symbol}
+
+flatten!(f!, out, x::Atom) = f!(out, x)
+flatten!(f!, out, xs)      = (for x in xs; flatten!(f!, out, x); end; out)
+
+nodes!(out, s::AbstractString) = push!(out, Node(TEXT, s))
+nodes!(out, n::Node)           = push!(out, n)
+
+function attributes!(out, s::AbstractString)
+    class, id = IOBuffer(), IOBuffer()
+    for x in eachmatch(r"[#|\.]([\w\-]+)", s)
+        print(startswith(x.match, '.') ? class : id, x.captures[1], ' ')
+    end
+    position(class) === 0 || push!(out, :class => rstrip(takebuf_string(class)))
+    position(id)    === 0 || push!(out, :id    => rstrip(takebuf_string(id)))
+    return out
+end
+attributes!(out, s::Symbol) = push!(out, s => "")
+attributes!(out, p::Pair)   = push!(out, p)
+
+function Base.show(io::IO, n::Node)
+    if istext(n)
+        print(io, escapehtml(n.text))
+    else
+        print(io, '<', n.name)
+        for (name, value) in n.attr
+            print(io, ' ', name)
+            isempty(value) || print(io, '=', repr(value))
+        end
+        if isvoid(n)
+            print(io, "/>")
+        else
+            print(io, '>')
+            if israw(n)
+                isempty(n.data) || print(io, n.data[1].text)
+            else
+                for each in n.data
+                    show(io, each)
+                end
+            end
+            print(io, "</", n.name, '>')
+        end
+    end
+end
+
+isvoid(n::Node) = n.name in VOID_ELEMENTS
```

Comment:
Ah, I would have them before normally, so that if you read linearly you would know the definitions of everything. But I don't know which is better and consistency is the key, so if it is after, it is after.